### ISSUE TITLE
OMEMO in Conversations client doesn't works out of the box

### ISSUE THREAD

**Issue by imShara** [processone/ejabberd#2425]
> What version of ejabberd are you using?

Ejabberd  18.04-1

> What operating system (version) are you using?

Linux archlinux 4.16.9-1-ARCH

> How did you install ejabberd (source, package, distribution)?

Pacman repository 

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

OMEMO Encryption doesn't work out of the box in Conversations client. No key sharing, message can't be send.

Logcat error:

> I/conversations(30616): AxolotlService (***@***.***): Creating axolotl sessions if needed...
> D/conversations(30616): AxolotlService (***@***.***): Finding devices without session for admin@ovel.space
> W/conversations(30616): AxolotlService (***@***.***): Have no target devices in PEP!

Ejabberd config:

```yaml
###
###'              ejabberd configuration file
###
###

### The parameters used in this configuration file are explained in more detail
### in the ejabberd Installation and Operation Guide.
### Please consult the Guide in case of doubts, it is included with
### your copy of ejabberd, and is also available online at
### http://www.process-one.net/en/ejabberd/docs/

### The configuration file is written in YAML.
### Refer to http://en.wikipedia.org/wiki/YAML for the brief description.
### However, ejabberd treats different literals as different types:
###
### - unquoted or single-quoted strings. They are called "atoms".
###   Example: dog, 'Jupiter', '3.14159', YELLOW
###
### - numeric literals. Example: 3, -45.0, .0
###
### - quoted or folded strings.
###   Examples of quoted string: "Lizzard", "orange".
###   Example of folded string:
###   > Art thou not Romeo,
###     and a Montague?

###.  =======
###'  LOGGING

##
## loglevel: Verbosity of log files generated by ejabberd.
## 0: No ejabberd log at all (not recommended)
## 1: Critical
## 2: Error
## 3: Warning
## 4: Info
## 5: Debug
##
loglevel: 4

##
## rotation: Describe how to rotate logs. Either size and/or date can trigger
## log rotation. Setting count to N keeps N rotated logs. Setting count to 0
## does not disable rotation, it instead rotates the file and keeps no previous
## versions around. Setting size to X rotate log when it reaches X bytes.
## To disable rotation set the size to 0 and the date to ""
## Date syntax is taken from the syntax newsyslog uses in newsyslog.conf.
## Some examples:
##  $D0     rotate every night at midnight
##  $D23    rotate every day at 23:00 hr
##  $W0D23  rotate every week on Sunday at 23:00 hr
##  $W5D16  rotate every week on Friday at 16:00 hr
##  $M1D0   rotate on the first day of every month at midnight
##  $M5D6   rotate on every 5th day of the month at 6:00 hr
##
log_rotate_size: 10485760
log_rotate_date: ""
log_rotate_count: 1

##
## overload protection: If you want to limit the number of messages per second
## allowed from error_logger, which is a good idea if you want to avoid a flood
## of messages when system is overloaded, you can set a limit.
## 100 is ejabberd's default.
log_rate_limit: 100

##
## watchdog_admins: Only useful for developers: if an ejabberd process
## consumes a lot of memory, send live notifications to these XMPP
## accounts.
##
## watchdog_admins:
##   - "bob@example.com"

###.  ===============
###'  NODE PARAMETERS

##
## net_ticktime: Specifies net_kernel tick time in seconds. This options must have
## identical value on all nodes, and in most cases shouldn't be changed at all from
## default value.
##
## net_ticktime: 60

###.  ================
###'  SERVED HOSTNAMES

##
## hosts: Domains served by ejabberd.
## You can define one or several, for example:
## hosts:
##   - "example.net"
##   - "example.com"
##   - "example.org"
##
hosts:
  - "***.***"

##
## route_subdomains: Delegate subdomains to other XMPP servers.
## For example, if this ejabberd serves example.org and you want
## to allow communication with an XMPP server called im.example.org.
##
## route_subdomains: s2s

###.  ============
###'  Certificates

## List all available PEM files containing certificates for your domains,
## chains of certificates or certificate keys. Full chains will be built
## automatically by ejabberd.
##
certfiles:
  - "/etc/letsencrypt/live/***.***/*.pem"
##   - "/etc/letsencrypt/live/example.com/*.pem"
##
## If your system provides only a single CA file (CentOS/FreeBSD):
## ca_file: "/etc/ssl/certs/ca-bundle.pem"

###.  =================
###'  TLS configuration

## Note that the following configuration is the default
## configuration of the TLS driver, so you don't need to
## uncomment it.
##
## define_macro:
##   'TLS_CIPHERS': "HIGH:!aNULL:!eNULL:!3DES:@STRENGTH"
##   'TLS_OPTIONS':
##     - "no_sslv3"
##     - "cipher_server_preference"
##     - "no_compression"
##   'DH_FILE': "/path/to/dhparams.pem" # generated with: openssl dhparam -out dhparams.pem 2048
##
## c2s_dhfile: 'DH_FILE'
## s2s_dhfile: 'DH_FILE'
## c2s_ciphers: 'TLS_CIPHERS'
## s2s_ciphers: 'TLS_CIPHERS'
## c2s_protocol_options: 'TLS_OPTIONS'
## s2s_protocol_options: 'TLS_OPTIONS'

###.  ===============
###'  LISTENING PORTS

##
## listen: The ports ejabberd will listen on, which service each is handled
## by and what options to start it with.
##
listen:
  -
    port: 5222
    ip: "::"
    module: ejabberd_c2s
    ##
    ## If TLS is compiled in and you installed a SSL
    ## certificate, uncomment this line:
    ##
    ## starttls: true
    ##
    ## To enforce TLS encryption for client connections,
    ## use this instead of the "starttls" option:
    ##
    starttls_required: true
    ##
    ## Stream compression
    ##
    zlib: true
    ##
    max_stanza_size: 65536
    shaper: c2s_shaper
    access: c2s
  ##
  ## Direct-TLS for C2S (XEP-0368). A good practice is to forward
  ## traffic from port 443 to this port, possibly multiplexing it
  ## with HTTP using e.g. sslh [https://wiki.xmpp.org/web/Tech_pages/XEP-0368],
  ## so modern clients can bypass restrictive firewalls (in airports, hotels, etc.).
  ##
  ## -
  ##   port: 5223
  ##   ip: "::"
  ##   module: ejabberd_c2s
  ##   tls: true
  ##   max_stanza_size: 65536
  ##   shaper: c2s_shaper
  ##   access: c2s
  -
    port: 5269
    ip: "::"
    module: ejabberd_s2s_in
  -
    port: 5280
    ip: "::"
    module: ejabberd_http
    request_handlers:
      "/ws": ejabberd_http_ws
      "/bosh": mod_bosh
      "/api": mod_http_api
    ##  "/pub/archive": mod_http_fileserver
    web_admin: true
    ## register: true
    captcha: true

  ##
  ## ejabberd_service: Interact with external components (transports, ...)
  ##
  ## -
  ##   port: 8888
  ##   ip: "::"
  ##   module: ejabberd_service
  ##   access: all
  ##   shaper_rule: fast
  ##   ip: "127.0.0.1"
  ##   privilege_access:
  ##      roster: "both"
  ##      message: "outgoing"
  ##      presence: "roster"
  ##   delegations:
  ##      "urn:xmpp:mam:1":
  ##        filtering: ["node"]
  ##      "http://jabber.org/protocol/pubsub":
  ##        filtering: []
  ##   hosts:
  ##     "icq.example.org":
  ##       password: "secret"
  ##     "sms.example.org":
  ##       password: "secret"

  ##
  ## ejabberd_stun: Handles STUN Binding requests
  ##
  ## -
  ##   port: 3478
  ##   transport: udp
  ##   module: ejabberd_stun

  ##
  ## To handle XML-RPC requests that provide admin credentials:
  ##
  ## -
  ##   port: 4560
  ##   ip: "::"
  ##   module: ejabberd_xmlrpc
  ##   maxsessions: 10
  ##   timeout: 5000
  ##   access_commands:
  ##     admin:
  ##       commands: all
  ##       options: []

  ##
  ## To enable secure http upload
  ##
  ## -
  ##   port: 5444
  ##   ip: "::"
  ##   module: ejabberd_http
  ##   request_handlers:
  ##     "": mod_http_upload
  ##   tls: true
  ##   protocol_options: 'TLS_OPTIONS'
  ##   dhfile: 'DH_FILE'
  ##   ciphers: 'TLS_CIPHERS'

## Disabling digest-md5 SASL authentication. digest-md5 requires plain-text
## password storage (see auth_password_format option).
## disable_sasl_mechanisms: "digest-md5"

###.  ==================
###'  S2S GLOBAL OPTIONS

##
## s2s_use_starttls: Enable STARTTLS for S2S connections.
## Allowed values are: false, optional or required
## You must specify 'certfiles' option
##
s2s_use_starttls: optional

##
## S2S whitelist or blacklist
##
## Default s2s policy for undefined hosts.
##
## s2s_access: s2s

##
## Outgoing S2S options
##
## Preferred address families (which to try first) and connect timeout
## in seconds.
##
## outgoing_s2s_families:
##   - ipv4
##   - ipv6
## outgoing_s2s_timeout: 190

###.  ==============
###'  AUTHENTICATION

##
## auth_method: Method used to authenticate the users.
## The default method is the internal.
## If you want to use a different method,
## comment this line and enable the correct ones.
##
## auth_method: internal

##
## Store the plain passwords or hashed for SCRAM:
## auth_password_format: plain
## auth_password_format: scram
##
## Define the FQDN if ejabberd doesn't detect it:
## fqdn: "server3.example.com"

##
## Authentication using external script
## Make sure the script is executable by ejabberd.
##
## auth_method: external
## extauth_program: "/path/to/authentication/script"

##
## Authentication using SQL
## Remember to setup a database in the next section.
##
## auth_method: sql

##
## Authentication using PAM
##
## auth_method: pam
## pam_service: "pamservicename"

##
## Authentication using LDAP
##
## auth_method: ldap
##
## List of LDAP servers:
## ldap_servers:
##   - "localhost"
##
## Encryption of connection to LDAP servers:
## ldap_encrypt: none
## ldap_encrypt: tls
##
## Port to connect to on LDAP servers:
## ldap_port: 389
## ldap_port: 636
##
## LDAP manager:
## ldap_rootdn: "dc=example,dc=com"
##
## Password of LDAP manager:
## ldap_password: "******"
##
## Search base of LDAP directory:
## ldap_base: "dc=example,dc=com"
##
## LDAP attribute that holds user ID:
## ldap_uids:
##   - "mail": "%u@mail.example.org"
##
## LDAP filter:
## ldap_filter: "(objectClass=shadowAccount)"

##
## Anonymous login support:
##   auth_method: anonymous
##   anonymous_protocol: sasl_anon | login_anon | both
##   allow_multiple_connections: true | false
##
## host_config:
##   "public.example.org":
##     auth_method: anonymous
##     allow_multiple_connections: false
##     anonymous_protocol: sasl_anon
##
## To use both anonymous and internal authentication:
##
## host_config:
##   "public.example.org":
##     auth_method:
##       - internal
##       - anonymous

###.  ==============
###'  DATABASE SETUP

## ejabberd by default uses the internal Mnesia database,
## so you do not necessarily need this section.
## This section provides configuration examples in case
## you want to use other database backends.
## Please consult the ejabberd Guide for details on database creation.

default_db: sql

##
## MySQL server:
##
## sql_type: mysql
## sql_server: "server"
## sql_database: "database"
## sql_username: "username"
## sql_password: "password"
##
## If you want to specify the port:
## sql_port: 1234

##
## PostgreSQL server:
##
sql_type: pgsql
sql_server: "localhost"
sql_database: "ejabberd"
sql_username: "ejabberd"
sql_password: "***"

## If you want to specify the port:
## sql_port: 1234
##
## If you use PostgreSQL, have a large database, and need a
## faster but inexact replacement for "select count(*) from users"
##
## pgsql_users_number_estimate: true

##
## SQLite:
##
## sql_type: sqlite
## sql_database: "/path/to/database.db"

##
## ODBC compatible or MSSQL server:
##
## sql_type: odbc
## sql_server: "DSN=ejabberd;UID=ejabberd;PWD=ejabberd"

##
## Number of connections to open to the database for each virtual host
##
## sql_pool_size: 10

##
## Interval to make a dummy SQL request to keep the connections to the
## database alive. Specify in seconds: for example 28800 means 8 hours
##
## sql_keepalive_interval: undefined

##
## Use the new SQL schema
##
## new_sql_schema: true

###.  ===============
###'  TRAFFIC SHAPERS

shaper:
  ##
  ## The "normal" shaper limits traffic speed to 1000 B/s
  ##
  normal: 1000

  ##
  ## The "fast" shaper limits traffic speed to 50000 B/s
  ##
  fast: 50000

##
## This option specifies the maximum number of elements in the queue
## of the FSM. Refer to the documentation for details.
##
max_fsm_queue: 10000

###.   ====================
###'   ACCESS CONTROL LISTS
acl:
  ##
  ## The 'admin' ACL grants administrative privileges to XMPP accounts.
  ## You can put here as many accounts as you want.
  ##
  admin:
    user:
      - "admin@***.***"
  ##     - "aleksey@localhost"
  ##     - "ermine@example.org"
  ##
  ## Blocked users
  ##
  ## blocked:
  ##   user:
  ##     - "baduser@example.org"
  ##     - "test"

  ## Local users: don't modify this.
  ##
  local:
    user_regexp: ""

  ##
  ## More examples of ACLs
  ##
  ## jabberorg:
  ##   server:
  ##     - "jabber.org"
  ## aleksey:
  ##   user:
  ##     - "aleksey@jabber.ru"
  ## test:
  ##   user_regexp: "^test"
  ##   user_glob: "test*"

  ##
  ## Loopback network
  ##
  loopback:
    ip:
      - "127.0.0.0/8"
      - "::1/128"
      - "::FFFF:127.0.0.1/128"

  ##
  ## Bad XMPP servers
  ##
  ## bad_servers:
  ##   server:
  ##     - "xmpp.zombie.org"
  ##     - "xmpp.spam.com"

##
## Define specific ACLs in a virtual host.
##
## host_config:
##   "localhost":
##     acl:
##       admin:
##         user:
##           - "bob-local@localhost"

###.  ============
###'  SHAPER RULES

shaper_rules:
  ## Maximum number of simultaneous sessions allowed for a single user:
  max_user_sessions: 10
  ## Maximum number of offline messages that users can have:
  max_user_offline_messages:
    - 5000: admin
    - 100
  ## For C2S connections, all users except admins use the "normal" shaper
  c2s_shaper:
    - none: admin
    - normal
  ## All S2S connections use the "fast" shaper
  s2s_shaper: fast

###.  ============
###'  ACCESS RULES
access_rules:
  ## This rule allows access only for local users:
  local:
    - allow: local
  ## Only non-blocked users can use c2s connections:
  c2s:
    - deny: blocked
    - allow
  ## Only admins can send announcement messages:
  announce:
    - allow: admin
  ## Only admins can use the configuration interface:
  configure:
    - allow: admin
  ## Only accounts of the local ejabberd server can create rooms:
  muc_create:
    - allow: local
  ## Only accounts on the local ejabberd server can create Pubsub nodes:
  pubsub_createnode:
    - allow: local
  ## In-band registration allows registration of any possible username.
  ## To disable in-band registration, replace 'allow' with 'deny'.
  register:
    - allow
  ## Only allow to register from localhost
  trusted_network:
    - allow: loopback
  ## Do not establish S2S connections with bad servers
  ## If you enable this you also have to uncomment "s2s_access: s2s"
  ## s2s:
  ##   - deny:
  ##     - ip: "XXX.XXX.XXX.XXX/32"
  ##   - deny:
  ##     - ip: "XXX.XXX.XXX.XXX/32"
  ##   - allow

## ===============
## API PERMISSIONS
## ===============
##
## This section allows you to define who and using what method
## can execute commands offered by ejabberd.
##
## By default "console commands" section allow executing all commands
## issued using ejabberdctl command, and "admin access" section allows
## users in admin acl that connect from 127.0.0.1 to  execute all
## commands except start and stop with any available access method
## (ejabberdctl, http-api, xmlrpc depending what is enabled on server).
##
## If you remove "console commands" there will be one added by
## default allowing executing all commands, but if you just change
## permissions in it, version from config file will be used instead
## of default one.
##
api_permissions:
  "console commands":
    from:
      - ejabberd_ctl
    who: all
    what: "*"
  "admin access":
    who:
      - access:
          - allow:
            - acl: loopback
            - acl: admin
      - oauth:
        - scope: "ejabberd:admin"
        - access:
          - allow:
            - acl: loopback
            - acl: admin
    what:
      - "*"
      - "!stop"
      - "!start"
  "public commands":
    who:
      - ip: "127.0.0.1/8"
    what:
      - "status"
      - "connected_users_number"

## By default the frequency of account registrations from the same IP
## is limited to 1 account every 10 minutes. To disable, specify: infinity
## registration_timeout: 600
  
##
## Define specific Access Rules in a virtual host.
##
## host_config:
##   "localhost":
##     access:
##       c2s:
##         - allow: admin
##         - deny
##       register:
##         - deny

###.  ================
###'  DEFAULT LANGUAGE

##
## language: Default language used for server messages.
##
language: "en"

##
## Set a different default language in a virtual host.
##
## host_config:
##   "localhost":
##     language: "ru"

###.  =======
###'  CAPTCHA

##
## Full path to a script that generates the image.
##
## captcha_cmd: "/lib/ejabberd/priv/bin/captcha.sh"

##
## Host for the URL and port where ejabberd listens for CAPTCHA requests.
##
## captcha_host: "example.org:5280"

##
## Limit CAPTCHA calls per minute for JID/IP to avoid DoS.
##
## captcha_limit: 5

###.  ====
###'  ACME
##
## In order to use the acme certificate acquiring through "Let's Encrypt"
## an http listener has to be configured to listen to port 80 so that
## the authorization challenges posed by "Let's Encrypt" can be solved.
## 
## A simple way of doing this would be to add the following in the listening
## section and to configure port forwarding from 80 to 5280 either via NAT
## (for ipv4 only) or using frontends such as haproxy/nginx/sslh/etc.
##   - 
##    port: 5280
##    ip: "::"
##    module: ejabberd_http

acme:

   ## A contact mail that the ACME Certificate Authority can contact in case of
   ## an authorization issue, such as a server-initiated certificate revocation.
   ## It is not mandatory to provide an email address but it is highly suggested.
   contact: "mailto:example-admin@example.com"


   ## The ACME Certificate Authority URL.
   ## This could either be:
   ##   - https://acme-v01.api.letsencrypt.org - (Default) for the production CA
   ##   - https://acme-staging.api.letsencrypt.org - for the staging CA
   ##   - http://localhost:4000 - for a local version of the CA
   ca_url: "https://acme-v01.api.letsencrypt.org"

###.  =======
###'  MODULES

##
## Modules enabled in all ejabberd virtual hosts.
##
modules:
  mod_adhoc: {}
  mod_admin_extra: {}
  mod_announce: # recommends mod_adhoc
    access: announce
  mod_blocking: {} # requires mod_privacy
  mod_caps: {}
  mod_carboncopy: {}
  mod_client_state: {}
  mod_configure: {} # requires mod_adhoc
  ## mod_delegation: {} # for xep0356
  mod_disco: {}
  mod_echo: {}
  mod_irc: {}
  mod_bosh: {}
  ## mod_http_fileserver:
  ##   docroot: "/var/www"
  ##   accesslog: "/var/log/ejabberd/access.log"
  ## mod_http_upload:
  ##   # docroot: "@HOME@/upload"
  ##   put_url: "https://@HOST@:5444"
  ##   thumbnail: false # otherwise needs ejabberd to be compiled with libgd support
  ## mod_http_upload_quota:
  ##   max_days: 30
  mod_last: {}
  ## XEP-0313: Message Archive Management
  ## You might want to setup a SQL backend for MAM because the mnesia database is
  ## limited to 2GB which might be exceeded on large servers
  ## mod_mam: {} # for xep0313, mnesia is limited to 2GB, better use an SQL backend
  mod_muc:
    ## host: "conference.@HOST@"
    access:
      - allow
    access_admin:
      - allow: admin
    access_create: muc_create
    access_persistent: muc_create
  mod_muc_admin: {}
  ## mod_muc_log: {}
  ## mod_multicast: {}
  mod_offline:
    access_max_user_messages: max_user_offline_messages
  mod_ping: {}
  ## mod_pres_counter:
  ##   count: 5
  ##   interval: 60
  mod_privacy: {}
  mod_private: {}
  ## mod_proxy65: {}
  mod_pubsub:
    access_createnode: pubsub_createnode
    ## reduces resource comsumption, but XEP incompliant
    ignore_pep_from_offline: true
    ## XEP compliant, but increases resource comsumption
    ## ignore_pep_from_offline: false
    last_item_cache: false
    plugins:
      - "flat"
      - "hometree"
      - "pep" # pep requires mod_caps
    force_node_config:
      ## Avoid using OMEMO by default because it
      ## introduces a lot of hard-to-track problems
      "eu.siacs.conversations.axolotl.*":
        access_model: whitelist
      ## Avoid buggy clients to make their bookmarks public
      "storage:bookmarks":
        access_model: whitelist
  mod_push: {}
  mod_push_keepalive: {}
  ## mod_register:
    ##
    ## Protect In-Band account registrations with CAPTCHA.
    ##
    ##   captcha_protected: true
    ##
    ## Set the minimum informational entropy for passwords.
    ##
    ##   password_strength: 32
    ##
    ## After successful registration, the user receives
    ## a message with this subject and body.
    ##
    ##   welcome_message:
    ##     subject: "Welcome!"
    ##     body: |-
    ##       Hi.
    ##       Welcome to this XMPP server.
    ##
    ## When a user registers, send a notification to
    ## these XMPP accounts.
    ##
    ##   registration_watchers:
    ##     - "admin1@example.org"
    ##
    ## Only clients in the server machine can register accounts
    ##
    ##   ip_access: trusted_network
    ##
    ## Local c2s or remote s2s users cannot register accounts
    ##
    ##   access_from: deny
    ##   access: register
  mod_roster: {}
  mod_shared_roster: {}
  mod_stats: {}
  mod_time: {}
  mod_vcard:
    search: false
  mod_vcard_xupdate: {}
  mod_avatar: {}
  mod_version: {}
  mod_stream_mgmt: {}
  ##   Non-SASL Authentication (XEP-0078) is now disabled by default
  ##   because it's obsoleted and is used mostly by abandoned
  ##   client software
  ## mod_legacy_auth: {}
  ##   The module for S2S dialback (XEP-0220). Please note that you cannot
  ##   rely solely on dialback if you want to federate with other servers,
  ##   because a lot of servers have dialback disabled and instead rely on
  ##   PKIX authentication. Make sure you have proper certificates installed
  ##   and check your accessibility at https://check.messaging.one/
  mod_s2s_dialback: {}
  mod_http_api: {}
  mod_fail2ban: {}

##
## Enable modules with custom options in a specific virtual host
##
## host_config:
##   "localhost":
##     modules:
##       mod_echo:
##         host: "mirror.localhost"

##
## Enable modules management via ejabberdctl for installation and
## uninstallation of public/private contributed modules
## (enabled by default)
##

allow_contrib_modules: true

###.
###'
### Local Variables:
### mode: yaml
### End:
### vim: set filetype=yaml tabstop=8 foldmarker=###',###. foldmethod=marker:
```

**Comment by zinid** [comment]
`grep OMEMO ejabberd.yml` maybe?

**Comment by imShara** [comment]
@zinid,  yep. But it so unclear.

**Comment by aiomaster** [comment]
Can someone please clarify what this means: 

> it introduces a lot of hard-to-track problems

What kind of problems can occur? I've never heard of one.

**Comment by weiss** [comment]
> What kind of problems can occur?

Many clients don't support it; e.g., *none* of the existing web clients I'm aware of currently do. Other clients don't support it properly; e.g., not in group chat or not with non-contacts. Other clients require manual installation of some plugin. Other clients don't hide key verification from users (or even enforce verification), so some users will just give up because they have no clue what they're supposed to do; others will verify keys and thereby break communication with unverified devices, e.g. when the contact buys a new phone. Also, OMEMO relies on PEP and (to also work with non-contacts) on some optional PubSub feature which not all servers support (properly), so if such a server is involved, this can lead to additional breakage.

Each of these issues is a UX nightmare, IMO. Power users might be able to cope with some of them, but ordinary mortals will just give up.

**Comment by aiomaster** [comment]
Hello @weiss,
thanks for you thoughts. In my personal opinion I don't think it's that worse. More and more clients support OMEMO or plan to support it in the future: https://omemo.top/
The legacy OTR is dropped from more and more clients instead.
And not every Client is a UX nightmare. Conversations does it very well in my opinion.
Key verification is a central part in that protocols and if you don't do it or can't do it, you have just less security and you should know it. Security is never for free and users have to understand the implications if a tool claims that.

The thing I just don't understand is, why the example configuration of ejabberd explicitly disables it.
If someone don't want to use it, its fine, but it doesn't seem to break something for users if it is just not disallowed.
In other words: I don't understand the logic in: "The UX implementation in some clients is bad, so better block it per default for all users." Or do I miss something here?

For me it is not a big thing, cause I can simply comment it out, but I just want to understand the reasons.

**Comment by zinid** [comment]
> I don't understand the logic in: "The UX implementation in some clients is bad, so better block it per default for all users."

And what's hard to understand here? We don't want XMPP users to have bad experience.
Also note that you can always enable OMEMO, but you cannot disable it because this will be treated as a downgrade attack by a client.

**Comment by aiomaster** [comment]
The funny thing is: I switched from prosody to ejabberd and my users (~50) had a bad experience cause OMEMO wasn't working out of the box and I had to find these lines to comment it out and make it work again.
I really don't get, why the server software should take care of the UX - this should be the responsibility of the clients. The server should just work and don't block something, cause it is thinking, that the clients are bad.
But hey, to make a constructive suggestion: You can add to the comment "If you want OMEMO to work properly, than you should comment out the following lines".
Thanks for your replies!

**Comment by aiomaster** [comment]
Wow this was fast. I still not agree that OMEMO should be blocked by default, but the new comment helps to know, what to do to enable it.
Thanks all for your statements and thanks to @imShara for the commit.

**Comment by weiss** [comment]
> Key verification is a central part in that protocols

Yes, if you use XMPP for highly sensitive communication (which I'm not sure is a good idea) and fear being subject of a man-in-the-middle tampering with your communication, then key verification makes sense.  So if clients want to support this use-case, they should offer verification.  But IMO this should be well hidden in some expert settings.  Normal people who just want to use XMPP for day-to-day communication just *won't* be able to cope with it.  As soon as the contact of a ChatSecure user buys a new phone, messages sent to him are silently lost and the ChatSecure user will have no clue how to fix it without asking some geek.

> Security is never for free and users have to understand the implications if a tool claims that.

The users I have in mind won't care at all about what you think they should understand.  They'll switch back to WhatsApp because our crap just doesn't work.

**Comment by arendtio** [comment]
@weiss You do know that conversations switched to TOFU a few months back? Since then it is very easy to use. Probably even too easy as I am not sure how secure that model is now.

Nevertheless, I just had a user walk over to me and ask why she couldn't send a message. The reason was simple: Today, I migrated my old config to the current example config and saw that comment and thought 'oh cool, I don't even have to remove the '#' they seem to support OMEMO out of the box'. I didn't notice that you boycott OMEMO out of the box :-1: 

Regarding hard to use: I have a few 50+ Users on my server which are neither hardcore XMPP users nor trained in IT personnel and yet they somehow manage to use OMEMO on a daily basis. In my **opinion**, that whole paragraph should be removed from the example config as it does no good there anyway.

**Comment by arendtio** [comment]
@licaon-kter sure, I am responsible for what I deploy, but that is not the point. I found it quite counterintuitive and I find the reasoning behind it wrong. I mean, that default config explicitly disables the most advanced end-to-end encryption we have in XMPP...

**Comment by licaon-kter** [comment]
@arendtio ...supported fully by how many clients exactly? Right, advanced, yes, trouble free on Conversations only...

**Comment by zinid** [comment]
Guys, the issue tracker is not the place for a drama, please chill out or I will have to mute the thread.

**Comment by weiss** [comment]
@arendtio 
> You do know that conversations switched to TOFU a few months back? Since then it is very easy to use.

You mean [Bilnd Trust Before Verification (BTBV)][1], which is a bit older.  BTBV works as long as you (1) don't use the (now prominently exposed) barcode scanner for adding contacts and (2) everyone uses Conversations.  However, ejabberd is an XMPP server, not a Conversations server.

[1]: https://gultsch.de/trust.html

**Comment by arendtio** [comment]
@weiss yes, I mixed up TOFU and BTBV, but that shouldn't make a big difference.

Nevertheless, I think I made my point and I would like to respect the warning of @zinid

**Comment by licaon-kter** [comment]
@paradisaeidae config?

**Comment by paradisaeidae** [comment]
Which section of ejabberd.yml is relevant?

**Comment by zinid** [comment]
@paradisaeidae mod_pubsub's options.

**Comment by paradisaeidae** [comment]
Thanks Zinid, will check that section tomorrow.

**Comment by paradisaeidae** [comment]
Thanks Zinid, 
Not sure what to change:

    mod_pubsub:
      access_createnode: pubsub_createnode
      ignore_pep_from_offline: false
      last_item_cache: false
      max_items_node: 1000000
    plugins: 
        - "flat"
        - "hometree"
        - "pep"
      force_node_config:      ## Avoid using OMEMO by default because it introduces a lot of hard-to-track problems
      "eu.siacs.conversations.axolotl.*":
          access_model: whitelist       ## Avoid buggy clients to make their bookmarks public
        "storage:bookmarks":
          access_model: whitelist

**Comment by licaon-kter** [comment]
@paradisaeidae Look at the same section in the updated example config: https://github.com/processone/ejabberd/blob/master/ejabberd.yml.example

**Comment by zinid** [comment]
Seems like there is also a mess with indentation.

**Comment by paradisaeidae** [comment]
Changed to this: Found the backtick+yaml for markdown technique!
```yaml
  mod_pubsub:
    access_createnode: pubsub_createnode
    ignore_pep_from_offline: false
    last_item_cache: false
    max_items_node: 1000000
    plugins: 
      - "flat"
      - "hometree"
      - "pep"
    force_node_config:      ## Avoid using OMEMO by default because it introduces a lot of hard-to-track problems
      "eu.siacs.conversations.axolotl.*":
        access_model: open       ## Avoid buggy clients to make their bookmarks public
      "storage:bookmarks":
        access_model: whitelist
```

**Comment by licaon-kter** [comment]
Those comments are kinda on the wrong rows, it doesn't matter for ejabberd but you might find them confusing in the future

**Comment by paradisaeidae** [comment]
Hi, OMEMO is the adventure!
Adjusted config to be pubsub open.
Now is see this in debug:
```
    sub_els =
        [#pubsub{
             subscriptions = undefined,subscription = undefined,
             affiliations = undefined,publish = undefined,
             publish_options = undefined,subscribe = undefined,
             unsubscribe = undefined,options = undefined,
             items =
                 #ps_items{
                     xmlns = <<>>,
                     node = <<"eu.siacs.conversations.axolotl.devicelist">>,
                     items = [],max_items = undefined,subid = <<>>,
                     retract = undefined},
             retract = undefined,create = undefined,configure = undefined,
             default = undefined,delete = undefined,purge = undefined,
             rsm = undefined},
         #stanza_error{
             type = cancel,code = 405,by = <<>>,reason = 'not-allowed',
             text = [],
             sub_els = [#ps_error{type = 'closed-node',feature = undefined}]}],
    meta = #{ip => {1,129,104,123}}}
```
Looks like this is from:
 https://github.com/processone/ejabberd/blob/d79ddd7b5c0a96c478df5a130cd62e43ef30c1db/src/mod_pubsub.erl#L1689

This is all I know today.

**Comment by licaon-kter** [comment]
@paradisaeidae Explain the log... what client... what did you do... etc...

**Comment by paradisaeidae** [comment]
My apologies,
My head is full, pretty close.
This is result of attempting to gain keys using gajim on Win10.
Gajim attempts to get keys reports: 'No devices found. Query in progress...'
Have restarted ejabberd, yep.
Previously on a fresh install of ejabberd keys are gained ok.
Changes of client seems hazardous.
Conversations reports fetching error: "There are no usable keys available for this contact.
Fetching new keys from the server has been unsuccessful. Maybe there is something wrong with your contact's server."
Would I have to delete the nodes for the new config to be applied?
I have tried doing this previously.
There is a mnesia table editor I used.
Though that was not effective, possibly due to my lack of understanding.

**Comment by paradisaeidae** [comment]
I am slowly understanding OMEMO and it's implementation mechanics.
Have had some days break.
Now it is clear that there are no keys for the client to read.
If omemo requires end-to-end key exchange then the peer target being offline will be an 'issue'.
So a more meaningful response to the request would be something which states this?

Interestingly the 'not-allowed' phrase is here: 
https://github.com/processone/ejabberd/search?q=not-allowed&unscoped_q=not-allowed
Which is in the muc code.
I am yet to understand why.

Could it be that ejabberd is not (able) not send a key request to the peer target device?
I don't see attempts to do this in the ejabberd debug log.
Though I do see the peer device (chatSecure) providing key in preKeyPublic preKeyId='37' type stanzas.
I see ejd supplying response stanzas with  type='result' to the chatSecure client.

More soon!

**Comment by tallero** [comment]
Just wanted to share the fact that on my ejabberd [server](https://arcipelago.ml) OMEMO key exchange between can be very slow (it requires from one minutes to three days for a client to see other contact's key the first time), while key exchanges between one own clients (2 gajim and 1 conversations for example) can take weeks or never happen in general.

Months ago both gajim AND conversations had a bug that made them not showing at all incoming encrypted messages they did not have key for (this did not happen for a while though), thus breaking sync and availability.

This tended to become very common when one used more than two devices to connect.

Also because on how OMEMO works (I believe, I am referring to the [100 keys](https://xmpp.org/extensions/xep-0384.html) rule) one has to wait for key syncing all over again when you do not use one of your device for long time.

Apart from slow key exchange, one-device accounts support is really good and configuration tends to be a one-timer.

This was to say that I can see why this is not enabled by default.
